File: Backtracking/N_Queen/CPP_N_Queen/N_Queen.cpp
#include <iostream>

#define MAX 14

using namespace std;

int n, ans;
bool height[MAX], down_top_diagonal[2 * MAX - 1], top_down_diagonal[2 * MAX - 1];

void n_queen(int i)
{
	if (i == n)
	{
		ans++;
		return;
	}

	for (int j = 0; j < n; j++)
	{
		if (height[j] || down_top_diagonal[i + j] || top_down_diagonal[i - j + n - 1])
			continue;

		height[j] = down_top_diagonal[i + j] = top_down_diagonal[i - j + n - 1] = true;
		n_queen(i + 1);
		height[j] = down_top_diagonal[i + j] = top_down_diagonal[i - j + n - 1] = false;
	}
}

int main()
{
	cin >> n;

	n_queen(0);

	cout << ans << "\n";

	return 0;
}

File: Backtracking/N_Queen/Go_N_Queen/N_Queen.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

const (
	max = 14
)

var (
	n               = 0
	ans             = 0
	height          [max]bool
	downTopDiagonal [2*max - 1]bool
	topDownDiagonal [2*max - 1]bool
)

func nQueen(i int) {
	if i == n {
		ans++
		return
	}

	for j := 0; j < n; j++ {
		if height[j] || downTopDiagonal[i+j] || topDownDiagonal[i-j+n-1] {
			continue
		}

		height[j] = true
		downTopDiagonal[i+j] = true
		topDownDiagonal[i-j+n-1] = true
		nQueen(i + 1)
		height[j] = false
		downTopDiagonal[i+j] = false
		topDownDiagonal[i-j+n-1] = false
	}
}

func main() {
	br := bufio.NewReader(os.Stdin)
	bw := bufio.NewWriter(os.Stdout)

	fmt.Fscanln(br, &n)

	nQueen(0)

	fmt.Fprintln(bw, ans)
	bw.Flush()
}
